java/vector/src/main/java/org/apache/arrow/vector/ipc/message/MessageResult.java

```diff
@@ -0,0 +1,83 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.vector.ipc.message;
+
+import java.nio.ByteBuffer;
+
+import org.apache.arrow.flatbuf.Message;
+
+/**
+ * Class to hold resulting Message and message information when reading messages through
+ * MessageReader.
+ */
+public class MessageResult {
+
+  /**
+   * Construct a container to hold a message result.
+   *
+   * @param messageLength the length of the message read in bytes
+   * @param messageBuffer contains the raw bytes of the message
+   * @param message the realized flatbuf Message
+   */
+  public MessageResult(int messageLength, ByteBuffer messageBuffer, Message message) {
```

Comment:
Why have message Result and MessageResult data as separate classes? Let's just collapse them into a single class.

The fact that a particular form of a message is backed by a ByteBuffer is probably an implementation detail. Why expose that in MessageResult?

Reply:
>Why have message Result and MessageResult data as separate classes?

This was so that I could return all of the information related to reading a flatbuf message, like this:
```java
public static MessageResult readMessage(ReadChannel in) throws IOException {
    ...
      messageLength = MessageSerializer.bytesToInt(buffer.array());

      // Length of 0 indicates end of stream
      if (messageLength != 0) {

        // Read the message into the buffer.
        messageBuffer = ByteBuffer.allocate(messageLength);
        if (in.readFully(messageBuffer) != messageLength) {
        ...
        // Load the message.
        message = Message.getRootAsMessage(messageBuffer);
      }
    }
    return new MessageResult(messageLength, messageBuffer, message);
  }
```
This way, the function produces a very clear output. Otherwise, if returning `MessageResultData<T>` it would need to be generic for no reason.